Feature/001 매장등록

TaehoLEEKR opened the pull request:
## 🎯 What is this PR
> 코드 변경 사유 및 목적 설명

## 📄 Changes Made
- 매장 등록 기능 구현 
- 매장Entity 생성
- GlobalException 구현
- GeoCoding API 로 주소를 통해 위도와 경도를 가져오는 기능 구현

## 📸 Screenshot
> 스크린 샷 첨부
![image](https://github.com/Siren-repo/Siren/assets/80689135/e21f0682-4d4a-4cbc-b393-5cd3b6c17efc)

## 🙋🏻‍ Review Point
- 리뷰어가 확인해야 할 사항 코멘트
- StoreService -> 입력 받은 role 을 관리 하는 부분
- domain 폴더 구조

## ✅ Test Checklist
- [x] API TEST ( PostMan)

## 🔗 Reference
Issue #7

Commit:
chore: 기본 디펜던시 추가 및 프로파일 그룹별 설정

TaehoLEEKR committed:
feat : 매장 등록 기능 구현 (#001);

body :
Store Entity 구현 Store 등록 기본 정보 엔티티 생성
GlobalException 구현 domain 폴더 구조에서 외부로 꺼내 글로벌로 관리
GeoCoding 기능 구현 Geocoding API 호출 을 위한 apiClass 는 utils에서 관리
StoreController 와 StoreService 구현 매장을 등록하는 POST(/register) 구현

현재 로그인 기능이 없어 클라이언트에서 role을 직접 입력 받아 데이터를 전달

TaehoLEEKR committed:
Merge branch 'develop' into feature/001-매장등록

yooyouny commented on src/main/java/com/devlop/siren/store/service/StoreService.java:
```diff
@@ -0,0 +1,51 @@
+
+    private final GeocodingApi geocodingApi;
+    private final StoreRepository storeRepository;
+    public boolean registerStore(String role, StoreRegisterDto storeRegisterDto) {
```
Post 요청을 처리하는 해당 메소드에 @Transactional이 필요하지 않을까요?

TaehoLEEKR replied:
아 트랜잭션을 뺴먹었었네요!

yooyouny commented on src/main/java/com/devlop/siren/store/controller/StoreController.java:
```diff
@@ -0,0 +1,21 @@
+public class StoreController {
+    private final StoreService storeService;
+    @PostMapping("/register")
+    public ResponseEntity<Boolean> registerStore(@RequestParam String role, @RequestBody StoreRegisterDto storeRegisterDto){
```
ResponseEntity보다는 공통된 응답객체를 만들어서 응답하는게 좋을것 같습니다

Ho-Tea replied:
공통적으로 반환되는 `ResponseDto`객체를 만들어서 `ResponseEntity`객체의 body값에 넣어 응답하는것을 말씀하시는 건가요?

yooyouny commented on src/main/java/com/devlop/siren/store/domain/Store.java:
```diff
@@ -0,0 +1,37 @@
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
```
@NoArgsConstructor 사용시 불완전한 객체 생성을 막기 위해 access를 설정하는 것이 좋을 것 같습니다

Ho-Tea replied:
저도 공부하면서 알았는데 `access = AccessLevel.PRIVATE` 로 설정하면 프록시 객체가 생성하지 못해 PROTECTED으로 설정해야한다고 합니다!

TaehoLEEKR replied:
저도 이 부분은 처음 알았네요

yooyouny commented on src/main/java/com/devlop/siren/store/domain/Store.java:
```diff
@@ -0,0 +1,37 @@
+
+@Entity
+@Getter
+@Setter
```
@Setter를 클래스 단에 붙이면 모든 필드에 접근해서 변경이 가능해지는데 변경이 필요한 속성에만 따로 메소드를 만드는게 안전하지 않을까요? 
평소에 @Setter, @AllArgsConstuctor, @Data등은 사용을 지양하고 @Builder도 클래스가 아닌 생성자 메소드 위에만 붙이는 편인데 엔티티 클래스에 붙이는 롬복 어노테이션들에 대해 다른 분들 의견은 어떤지 궁금합니다

Ho-Tea replied:
[Entity의 lombok](https://velog.io/@langoustine/setter-%EC%A7%80%EC%96%91-%EC%9D%B4%EC%9C%A0) 이 게시글에도 잘 나와있듯이 Setter를 사용하는 것은 여러 곳에서 Entity객체에 접근할 수 있으므로 일관성이 무너질 것 같습니다
유연님이 말씀하신 @Builder 도 클래스 딴이 아닌 생성자 메서드 위에만 붙이는 방식도 일관성을 유지하기 위함 이므로, Entity를 일관성을 깨뜨릴 수 있는 @Setter, @Data는 지양하고 @AllArgsConstructor 대신에 모든 필드를 받는 생성자를 생성하고 그 생성자 위에 @Builder를 다는 형식으로 엔티티의 일관성을 유지하는것이 더 좋다고 생각합니다

TaehoLEEKR replied:
그럼 저도 @Builder 는 생성하고자 하는 생정자 위에 생성하는 형식으로 바꾸고 @Setter , @AllArgsConstructor 를 지양 하는식으로 작성하겠습니다 👍

yooyouny commented on src/main/java/com/devlop/siren/store/domain/Store.java:
```diff
@@ -0,0 +1,37 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long storeId;
+    @Column(name = "storeName")
```
@Column의 name속성은 db에 저장되는 컬럼 이름이기 때문에 자바에서 사용하는 카멜케이스가 아닌 스네이크케이스(store_name)으로 작성해야 할 것 같습니다

yooyouny commented on src/main/java/com/devlop/siren/store/dto/StoreRegisterDto.java:
```diff
@@ -0,0 +1,35 @@
+import java.time.LocalDateTime;
+
+@Getter
+@Setter
```
사용할 롬복 어노테이션만 정의해주는게 좋을것 같습니다 다른곳에서 해당 dto의 set메소드를 사용하지는 않는것 같습니다

yooyouny commented on src/main/java/com/devlop/siren/store/service/StoreService.java:
```diff
@@ -0,0 +1,51 @@
+            throw new GlobalException(ErrorCode.NOT_AUTH_ROLE);
+        }
+
+        try {
```
해당 registerStore 메소드는 스토어에 저장하는 역할을 하는 메서드로 남겨놓고 지오코딩 관련 로직은 private 메서드로 따로 분리하는 등으로 처리하는게 가독성, 재사용성면에서 좋을 것 같습니다

yooyouny commented on src/main/resources/application.yml:
```diff
@@ -28,14 +46,72 @@ spring:
+      hibernate:
+        format_sql: 'true'
+
+google:
```
저희 공통적으로 사용하는 key값 같은 것은 common 프로파일에서 따로 관리하고 있어서 그 위에 작성해주시면 좋을 것 같습니다(위에 jwt key 작성되어있는 곳)

yooyouny commented on src/main/java/com/devlop/siren/store/utils/GeocodingApi.java:
```diff
@@ -0,0 +1,25 @@
+    public GeocodingResult[] geocodeAddress(String address) {
+        try {
+            return com.google.maps.GeocodingApi.geocode(context, address).await();
+        } catch (Exception e) {
```
캐치로 에러를 잡고 출력만 하지 아무 처리를 안하는 것 같은데 디버깅 하실때 작성한 코드일까요? 런타임 에러로 새로 던지는 등 에러처리가 필요할 것 같습니다

TaehoLEEKR replied:
geocode API 를 호출 하기 위해서 .await() 를 써야하는데 이 구문은 I/O, Runtime , API Exception 을 처리 해야하기 때문에  
    public GeocodingResult[] geocodeAddress(String address) throws IOException, InterruptedException, ApiException {
            return com.google.maps.GeocodingApi.geocode(context, address).await();
    }
    throws 를 추가 하여 에러를 처리하거나
    try-catch 를 사용하는 방법이라 try-catch Exception 으로 잡아놨습니다

yooyouny commented on src/main/java/com/devlop/siren/store/dto/StoreRegisterDto.java:
```diff
@@ -0,0 +1,35 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class StoreRegisterDto {
```
저는 db관련된 설정(unique 컬럼 설정 등)은 엔티티에서 처리하고 Dto에서 값 검증을 처리하는데 아직 dto나 엔티티에서 validation처리가 보이지 않는 것 같습니다 
String은 @NotBlank로 처리하거나 zipcode는 숫자5자리 제한 등등이 필요할것 같습니다.

yooyouny commented on src/main/java/com/devlop/siren/exception/ErrorCode.java:
```diff
@@ -0,0 +1,15 @@
+package com.devlop.siren.exception;
```
이번에 올리신 exception관련 코드들은 공통내용이라 com/develop/siren/global/exception 경로에 있는게 맞을 것 같습니다

yooyouny commented on src/main/java/com/devlop/siren/store/service/StoreService.java:
```diff
@@ -0,0 +1,51 @@
+package com.devlop.siren.store.service;
```
도메인 프로젝트 구조 글을 참고하면 service 패키지 명을 application으로 수정해야 할 것 같습니다

Ho-Tea replied:
이부분도 Controller와 마찬가지로 Service 패키지명으로 유지해야한다는 입장인데 어떻게 생각하실까요?

yooyouny commented on src/main/java/com/devlop/siren/store/controller/StoreController.java:
```diff
@@ -0,0 +1,21 @@
+package com.devlop.siren.store.controller;
```
도메인 프로젝트 구조 글을 참고하면 controller 패키지 명을 api로 수정해야 할 것 같습니다

Ho-Tea replied:
저희가 참고한 도메인프로젝트 구조를 제외하고 다른 도메인 프로젝트 구조를 참고해 봤을때 Controller로 구성한 구조도 많았고,
저 또한 MVC패턴에 익숙해져 있어서 Controller라는 패키지명이 더 와닿는데 Controller로 유지하는건 어떨까요?

TaehoLEEKR replied:
저도 MVC 패턴에 익숙해져있어서 Controller/Service 가 익숙합니다

yooyouny reviewed (commented):
일단 보이는 내용으로 코멘트 달아놨는데 미처 발견하지 못한 내용이 있을 수도 있습니다 🥲
리뷰 내용을 크게 나눠보면 어노테이션, 패키지구조 같은데 해당 주제에 대해 같이 얘기해보면 좋을것 같습니다.

그리고 PR을 올리면 github actions로 테스트 자동화를 수행하도록 한 것인데 해당 코드는 api까지 만들어져있지만 테스트 코드가 전혀 없는 상태라 해당 기능을 제대로 수행하는지, 예상했던 예외가 잘 던져지는지 리뷰이의 입장에서는 확신을 할 수가 없습니다.. 아직 유저 인증부분이 만들어지지 않은터라 컨트롤러 유닛테스트는 제외 하더라도 develop에 머지하기 위해서는 서비스, 레파지토리 유닛 테스트는 필요하지 않을까 싶습니다.

Ho-Tea commented on src/main/java/com/devlop/siren/store/domain/Store.java:
```diff
@@ -0,0 +1,37 @@
+    private String storePhone;
+    @Column(name = "city",columnDefinition = "NVARCHAR(255) NOT NULL" )
+    private String city;
+    @Column(name = "street",columnDefinition = "NVARCHAR(255) NOT NULL" )
```
`columnDefinition` 으로 NOT NULL 옵션을 주기보다 `@NotNull` 애노테이션을 사용하는 것이 가독성 측면에서 더 좋다고 생각하는데 어떻게 생각하시나요?

Ho-Tea replied:
추가적으로, Entity에도 DTO에서 검증하는것 말고도 추가적으로 검증로직을 추가(ex: 길이제한 `@Size`)하는것에 대해서 고민인데 다른분들은 어떻게 생각하실까요?

TaehoLEEKR replied:
제가 알기론 columnDefinition 은 데이터베이스의 컬럼 타입의 제약 사항을 걸어주는 부분이고 @NotNull 은 DTO 쪽에 맞다고 생각이 들어요 만약에 @Size 처럼 검증 로직이 필요하다면 Column은 그대로 두고 @Length 라는 어노테이션으로 길이를 제한하는 부분이 저는 좋다고 생각이들어요

Ho-Tea commented on src/main/java/com/devlop/siren/store/repository/StoreRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.Optional;
+
+@Repository
```
스프링부트가 JPARepository를  implements하는 리파지토리들은 자동으로 `@Repository`로 등록하는데 혹시 중복사용하신 다른 이유가 있을까요?

TaehoLEEKR replied:
아 이부분은 제 코드스타일이긴 한데 @Service / @RestContreollr / @Component / @Configuration / @Repository 를 명시해주는 편입니다 현재 Repository에 있는어노테이션은 이미 implements 해서 생략 가능한 부분입니다.

Ho-Tea commented on src/main/java/com/devlop/siren/exception/GlobalException.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+
+@Getter
+public class GlobalException extends RuntimeException {
```
단순히 예외 메시지를 전달하는 목적이라면 커스텀 예외로 정의하지말고 표준예외로 처리하는건 어떨까요? 
지금은 예외사항이 얼마 없어 괜찮지만 추후 프로젝트가 커지면서 여러 예외사항들이 커스텀 예외로 만들어 지면 그것또한 유지보수하기 힘들 것 같다는 생각입니다

Ho-Tea reviewed (commented):
유연님이 리뷰하신 부분에서 공감되는 부분도있고 다르게 생각되는 부분도 있어 코멘트 남겨놓았습니다!
전체적으로 공감되는 부분은 검증로직을 추가한다거나 Entity의 일관성과 같은 비즈니스로직 부분에서 많이 보였고, 말씀하신 패키지구조같은 부분에서 저의 견해와 다르게 생각되는 부분이 있었습니다
유연님이 앞서 코드리뷰를 잘해주셔가지고 결정적인 코드의 문제점(?)과 같은 부분은 찾지 못했고 제가 궁금한 사항들 또한 코멘트 남겨놓았습니다

또한, Jacoco가 테스트 커버리지의 결과를 측정하는 라이브러리인 만큼 어느정도의 테스트 코드는 필요하다는 생각입니다

TaehoLEEKR committed:
style : PR 리뷰 피드백 변경(#6);

body :
Errorcode 폴더 위치 변경 > global
Entity Setter , AllArgsConstructor , 사용 지양
NoArgsConstructor access 제한
DTO validation 설정
Service @Transactional 추가
